[JENKINS-26781] Revise #1443 to support descriptors with ID != clazz.name

ndeloof opened the pull request:
[JENKINS-26781](https://issues.jenkins-ci.org/browse/JENKINS-26781)

so descriptor can be retrieved by ID, which _may_ not be clazz.name

@reviewbybees

ndeloof committed:
[JENKINS-26781] lookup descriptor by ID, then by class if explicitly set
to match some specific use cases.

jglick commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -222,7 +222,7 @@ public Descriptor getItemTypeDescriptorOrDie() {
      * Help file redirect, keyed by the field name to the path.
      *
-     * @see #getHelpFile(String) 
+     * @see #getHelpFile(String)
```
Please do not make insignificant whitespace changes in PRs.

jglick commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -918,6 +918,7 @@ private URL getStaticHelpUrl(Klass<?> c, String suffix) {
                 if (d != null) {
                     items.add(d.newInstance(req, jo));
                 }
+                LOGGER.warning("Received unexpected formData for descriptor " + kind);
```
Should be in an `else`-clause.

ndeloof replied:
feeling stupid

jglick commented on core/src/main/resources/lib/form/class-entry.jelly:
```diff
@@ -34,7 +34,7 @@ THE SOFTWARE.
     </st:attribute>
   </st:documentation>
-  <j:set var="clazz" value="${attrs.clazz ?: attrs.descriptor.clazz.name}" />
+  <j:set var="clazz" value="${attrs.clazz ?: attrs.descriptor.id}" />
```
Should adjust documentation for `clazz` to emphasize that it should be `Descriptor.id` more generally.

ndeloof commented:
investigating on strange regression found by test harness ...

ndeloof commented:
this test case demonstrate some code still relies on stapler-class

ndeloof commented:
commented on https://github.com/jenkinsci/jenkins/pull/1443#issuecomment-72935118
IHMO there's an implementation issue using stapler "$class" parameter to pass an arbitrary ID

ndeloof commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -909,14 +909,19 @@ private URL getStaticHelpUrl(Klass<?> c, String suffix) {
+                Descriptor<T> d = null;
+                String kind = jo.optString("kind", null);
+                if (kind != null) {
+                    d = find(descriptors, kind);
```
I wonder it would be cleaner to introduce new "FindById" and "findByClassName" methods and explicitly invoke one of them. find would then use both and marked as deprecated.

jglick replied:
Might made things clearer.

ndeloof committed:
deprecated find(string) and introduce explicit methods
to lookup descriptor by id or by Describable class name

jglick commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -925,22 +930,40 @@ private URL getStaticHelpUrl(Klass<?> c, String suffix) {
+
+    /**
+     * Finds a descriptor from a collection by its class name.
+     * @deprecated Since we introduced {@link }Descriptor#getId()}, it is a preferred method of identifying descriptor by a string.
```
Extraneous `}`.

jglick commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -925,22 +930,40 @@ private URL getStaticHelpUrl(Klass<?> c, String suffix) {
     }
 
+    /**
+     * Finds a descriptor from a collection by its class name.
```
Documentation should clarify: class name _or_ ID.

ndeloof commented:
@mattmoor being author for the initial PR your review would be appreciated

mattmoor commented:
Can you please add a test for this "supported" behavior?  It clearly isn't covered today.

IMO the evolution of this should be that "$class" and "kind" are mutually exclusive.  Emit "kind" iff it isn't clazz.name, otherwise emit it as "$class".

My $0.02

jglick commented:
Agreed that test coverage for this needs to be added to core. IIUC @kohsuke introduced an overridable `id` to make things like `cloudbees-template` possible, but did not cover this mode in a core test case.

Using `$class` when `d.id == d.clazz.name` and `kind` otherwise makes some sense to me. In the latter case the default implementation of `newInstance` would not work; the descriptor impl is expected to have custom handling.

ndeloof commented:
I honestly don't understand descriptor-based databinding enough so I can propose a unit test to cover this.

ndeloof commented:
We can't define $class and kind to be mutually exclusive, as DownloadFromUrlInstaller.DescriptorImpl do override getId, so this would enforce ALL tool installer plugins to override newInstance.

jglick commented:
Only `DownloadFromUrlInstaller` subclasses would need to override `newInstance` in the descriptor, not all `ToolInstaller`s.

Of course in this case I question why the `getId` override is needed to begin with. It seems gratuitous. Simpler and safer would be to do the character substitution in the call to `createDownloadable`.

ndeloof committed:
make $class and kind mutually exclusive
Use kind when descriptor.id != descriptor.clazz.name and $clazz otherwise.

stephenc commented:
Where are we on this? it seems to me like this pull request highlights that the choice of `$class` and the "standardization" around it was incorrect and probably needs to be "reverted" with a different approach taken

jglick commented:
> Where are we on this?

I have not personally looked into it deeply enough to take a stance.

jglick committed:
[JENKINS-26781] Reproduced problem in test.
Besides failing in master, it also fails in 4e0af43 (merge of #1443), but passes in the 1.598-SNAPSHOT parent 239caf8.

jglick committed:
Merge remote-tracking branch 'origin/JENKINS-26781' into JENKINS-26781

jglick committed:
Now it passes.

jglick committed:
Added @since.

jglick committed:
Improving Javadoc.

jglick committed:
findByClassName should not be advertised to outside code.

jglick committed:
Noting deprecations more widely.

jglick commented:
GitHub seems confused. [Running diff](https://github.com/jenkinsci/jenkins/compare/master...JENKINS-26781)

jglick commented:
> Can you please add a test for this "supported" behavior? It clearly isn't covered today.

@mattmoor agreed, and added such a test belatedly.

kohsuke commented:
So IIUC, in the new version `kind` and `$class` is mutually exclusive, and what you get depends on whether `Descriptor.id` and `Descriptor.clazz` are the same?

I thought the idea behind `$class` was that we are just changing the keyword from `kind`/`stapler-class` to `$class` without its meaning --- namely `$class` is always compared against `Descriptor.id` despite its name, and so it's usually a class name but not all the time.

If so, shouldn't the change be editing `class-entry.jelly` so that it bakes `descriptor.id` as `$class` and not `descriptor.clazz.name`?

jglick commented:
> in the new version `kind` and `$class` is mutually exclusive, and what you get depends on whether `Descriptor.id` and `Descriptor.clazz` are the same?

Right. Typically `$class` is used, but `kind` is used instead when a custom ID is in play. (For example, when saving a job using builder and/or publisher templates.)

> so it's usually a class name but not all the time

I would find that confusing. The very term “class” indicates that it _is_ a class name.

FYI, when used from `DescribableHelper` in Workflow, `$class` is expected to be a `Describable` `Class.name`/`.simpleName`, and that concrete class is instantiated according to `@DataBoundConstructor`/`@DataBoundSetter`. There is no other real option—`Descriptor.newInstance(StaplerRequest, JSONObject)` cannot be used in this context, so the `Descriptor` is useless as well.

kohsuke committed:
$class is not legacy, stapler-class is, so making sure it's not getting mixed up.

kohsuke committed:
Doc improvements

kohsuke commented:
OK. I added doc to capture that.

kohsuke commented:
:+1:

jglick committed:
Merge branch 'master' into JENKINS-26781 and fixing @since TODO.

jglick committed:
Merge branch 'JENKINS-26781' of github.com:jenkinsci/jenkins into JENKINS-26781

jglick committed:
Reverting whitespace change.

jglick committed:
Merge branch 'master' into JENKINS-26781

jglick commented:
FTR: the PR builder was down, then the build of 2dcb6d3 timed out, then the build of f4ddac1 got eight failures but all from

    java.lang.NoClassDefFoundError: Could not initialize class com.cloudbees.plugins.credentials.SystemCredentialsProvider$ProviderImpl

which is a “well-known” random failure (CC @stephenc). I reran those eight tests locally in this branch and they all pass. So merging.

daniel-beck commented:
Appears to have broken Copy Artifact: [JENKINS-28011](https://issues.jenkins-ci.org/browse/JENKINS-28011)

daniel-beck commented on core/src/main/java/hudson/tools/DownloadFromUrlInstaller.java:
```diff
@@ -122,15 +122,16 @@ protected DescriptorImpl() {
 
         protected Downloadable createDownloadable() {
-            return new Downloadable(getId());
+            return new Downloadable(getDownloadableId());
```
This broke the tool installer for NodeJS Plugin: [JENKINS-28093](https://issues.jenkins-ci.org/browse/JENKINS-28093)

daniel-beck commented:
Appears to have broken more than just the tool installer in NodeJS Plugin: [JENKINS-28114](https://issues.jenkins-ci.org/browse/JENKINS-28114)

daniel-beck commented:
May have broken Performance Plugin: [JENKINS-28110](https://issues.jenkins-ci.org/browse/JENKINS-28110)

daniel-beck commented:
With plugins seemingly breaking left and right I wonder whether this should be rolled back.

jglick commented:
I plan to take a look at these and see if they are really three separate issues or if there is some duplication.

ikedam commented on core/src/main/resources/lib/form/class-entry.jelly:
```diff
@@ -36,12 +46,14 @@ THE SOFTWARE.
+    <j:choose>
+      <j:when test="${descriptor != null and descriptor.id != clazz}">
+        <input type="hidden" name="kind" value="${attrs.descriptor.id}" />
+      </j:when>
```
[JENKINS-28114 NodeJsInstaller exception](https://issues.jenkins-ci.org/browse/JENKINS-28114) is caused for here.
* `NodeJsInstaller` overrides `Descriptor#getId` (https://github.com/jenkinsci/nodejs-plugin/blob/nodejs-0.2.1/src/main/java/jenkins/plugins/nodejs/tools/NodeJSInstaller.java#L383). This results no `$class` nor `stapler-class` as `descriptor.id != clazz`
* An exception occurs when the data binding is handled by stapler, as stapler never recognizes `kind`.

ikedam replied:
Same for [JENKINS-28110 Performance plugin problem](https://issues.jenkins-ci.org/browse/JENKINS-28110).
performance-plugin ovierrides `Descriptor#getId` for almost all its descriptors (overriding in the base class https://github.com/jenkinsci/performance-plugin/blob/performance-1.12/src/main/java/hudson/plugins/performance/PerformanceReportParserDescriptor.java#L16).

jglick commented:
@ikedam thanks for tips, that should help me get started faster.

jglick commented:
_Why_ the `performance` plugin overrides `getId` is beyond me. This seems be used only by a `getById` method, which is unused. https://github.com/jenkinsci/performance-plugin/commit/d93fb0c2df9c0d8701eca48d501a74d40d05546e At any rate, it serves as a demo of a plugin which does something weird and which must be supported.

kohsuke commented:
Since Jesse made the point that I am responsible for the problematic commit in the performance plugin, I feel the need to defend myself. AFAICT, the performance plugin had `getId()` method first, then the core added `Descriptor.getId()` later.

jglick commented:
I see, did not realize this was a recentish addition.

(A situation where it would be nice to make javac upgrade the warning about missing `@Override` to an error.)
